Sources/GRPCHTTP2Core/Client/Connection/LoadBalancers/RoundRobinLoadBalancer.swift

```diff
@@ -0,0 +1,766 @@
+          let subchannel = makeSubchannel(key.endpoint, SubchannelID())
+          subchannels.append(subchannel)
+          self.subchannels[key] = SubchannelState(subchannel: subchannel)
+        }
+
+        return subchannels
+      }
+    }
+
+    struct Closing {
+      enum Reason: Sendable, Hashable {
+        case goAway
+        case user
+      }
+
+      var reason: Reason
+      var parkedSubchannels: [EndpointKey: Subchannel]
+
+      mutating func updateConnectivityState(_ state: ConnectivityState, key: EndpointKey) -> Bool {
+        switch state {
+        case .idle, .connecting, .ready, .transientFailure:
+          ()
+        case .shutdown:
+          self.parkedSubchannels.removeValue(forKey: key)
+        }
+
+        return self.parkedSubchannels.isEmpty
+      }
+    }
+
+    struct SubchannelState {
+      var subchannel: Subchannel
+      var state: ConnectivityState
+      var markedForRemoval: Bool
+
+      init(subchannel: Subchannel) {
+        self.subchannel = subchannel
+        self.state = .idle
+        self.markedForRemoval = false
+      }
+
+      mutating func updateState(_ newState: ConnectivityState) -> Bool {
+        // The transition from transient failure to connecting is ignored.
+        //
+        // See: https://github.com/grpc/grpc/blob/master/doc/load-balancing.md
+        if self.state == .transientFailure, newState == .connecting {
+          return false
+        }
+
+        let oldState = self.state
+        self.state = newState
+        return oldState != newState
+      }
+
+      mutating func markForRemoval() {
+        self.markedForRemoval = true
+      }
+    }
+
+    struct Picker {
+      private var subchannels: [Subchannel]
+      private var index: Int
+
+      init?(subchannels: [Subchannel]) {
+        if subchannels.isEmpty { return nil }
+
+        self.subchannels = subchannels
+        self.index = (0 ..< subchannels.count).randomElement()!
+      }
+
+      mutating func pick() -> Subchannel {
+        defer {
+          self.index = (self.index + 1) % self.subchannels.count
+        }
+        return self.subchannels[self.index]
+      }
+    }
+
+    mutating func updateSubchannels(
+      newEndpoints: Set<EndpointKey>,
+      makeSubchannel: (_ endpoint: Endpoint, _ id: SubchannelID) -> Subchannel
+    ) -> (run: [Subchannel], close: [Subchannel], newState: ConnectivityState?) {
+      switch self {
+      case .active(var state):
+        let existingEndpoints = Set(state.subchannels.keys)
+        let keysToAdd = newEndpoints.subtracting(existingEndpoints)
+        let keysToRemove = existingEndpoints.subtracting(newEndpoints)
+
+        if keysToRemove.isEmpty && keysToAdd.isEmpty {
+          // Nothing to do.
+          return (run: [], close: [], newState: nil)
+        }
+
+        // The load balancer should keep subchannels to remove in service until they a new subchannel
+        // can replace it so that requests can continue to be served.
```

Comment:
Nit, something like:
```suggestion
        // The load balancer should keep subchannels to remove in service until a new subchannel
        // can replace each of them, so that requests can continue to be served.
```